tests/openwisp2/sample_firmware_upgrader/tests.py

```diff
@@ -1,108 +1,55 @@
     build_list_url = reverse(f'admin:{app_label}_build_changelist')
-    device_firmware_model = DeviceFirmware
-    upgrade_operation_model = UpgradeOperation
-    firmware_image_model = FirmwareImage
-    build_model = Build
-    category_model = Category
 
 
-@skipUnless(
-    os.environ.get('SAMPLE_APP', False),
-    'Running tests on standard openwisp_firmware_upgrader models',
-)
-class TestModels(BaseTestModels, TestUpgraderMixin, TestCase):
-    app_name = 'openwisp2.sample_firmware_upgrader'
-    device_firmware_model = DeviceFirmware
-    upgrade_operation_model = UpgradeOperation
-    batch_upgrade_operation_model = BatchUpgradeOperation
-    firmware_image_model = FirmwareImage
-    build_model = Build
-    category_model = Category
+class TestModels(BaseTestModels):
+    app_label = 'openwisp2.sample_firmware_upgrader'
 
 
-@skipUnless(
-    os.environ.get('SAMPLE_APP', False),
-    'Running tests on standard openwisp_firmware_upgrader models',
-)
-class TestModelsTransaction(
-    BaseTestModelsTransaction, TestUpgraderMixin, TransactionTestCase
-):
-    device_firmware_model = DeviceFirmware
-    upgrade_operation_model = UpgradeOperation
-    batch_upgrade_operation_model = BatchUpgradeOperation
-    firmware_image_model = FirmwareImage
-    build_model = Build
-    category_model = Category
+class TestModelsTransaction(BaseTestModelsTransaction):
+    pass
 
 
-@skipUnless(
-    os.environ.get('SAMPLE_APP', False),
-    'Running tests on standard openwisp_firmware_upgrader models',
-)
-class TestOpenwrtUpgrader(BaseTestOpenwrtUpgrader, TransactionTestCase):
-    device_firmware_model = DeviceFirmware
-    upgrade_operation_model = UpgradeOperation
-    batch_upgrade_operation_model = BatchUpgradeOperation
-    firmware_image_model = FirmwareImage
-    build_model = Build
-    category_model = Category
+class TestOpenwrtUpgrader(BaseTestOpenwrtUpgrader):
+    pass
 
 
-@skipUnless(
-    os.environ.get('SAMPLE_APP', False),
-    'Running tests on standard openwisp_firmware_upgrader models',
-)
-class TestPrivateStorage(BaseTestPrivateStorage, TestUpgraderMixin, TestCase):
-    device_firmware_model = DeviceFirmware
-    upgrade_operation_model = UpgradeOperation
-    batch_upgrade_operation_model = BatchUpgradeOperation
-    firmware_image_model = FirmwareImage
-    build_model = Build
-    category_model = Category
+class TestPrivateStorage(BaseTestPrivateStorage):
+    pass
 
 
-@skipUnless(
-    os.environ.get('SAMPLE_APP', False),
-    'Running tests on standard openwisp_firmware_upgrader models',
-)
-class TestTasks(BaseTestTasks, TestUpgraderMixin, TransactionTestCase):
-    device_firmware_model = DeviceFirmware
-    upgrade_operation_model = UpgradeOperation
-    batch_upgrade_operation_model = BatchUpgradeOperation
-    firmware_image_model = FirmwareImage
-    build_model = Build
-    category_model = Category
+class TestTasks(BaseTestTasks):
+    pass
+
+
+# this is necessary to avoid excuting the base test suites
```

Comment:
This part looks very hacky. What if the base classes don't just inherit from TestCase?

E.g.:

```diff
# openwisp_firmware_upgrader/tests/test_models.py
-class TestModels(TestUpgraderMixin, TestCase):
+class TestModels(TestUpgraderMixin):
    app_label = 'openwisp_firmware_upgrader'
   ...
```

and then

```diff
# tests/openwisp2/sample_firmware_upgrader/tests.py
-class TestModels(BaseTestModels):
+class TestModels(BaseTestModels, TestCase):
    app_label = 'openwisp2.sample_firmware_upgrader'
```

Reply:
I agree is hacky.

The solution proposed would result in tests not being executed when runnning: `./runtests.py`
Only `SAMPLE_APP=1 ./runtests.py` would run actual tests.

The alternative would be:

```diff
# openwisp_firmware_upgrader/tests/test_models.py
-class TestModels(TestUpgraderMixin, TestCase):
+class BaseTestModels(TestUpgraderMixin):
    app_label = 'openwisp_firmware_upgrader'
   ...


+class TestModels(BaseTestModels):
+    pass
```

Which is somewhat less hacky but still weird, so I thought that if we have to have weird code (since there's no way around it), having it in the sample app is better since we deal with it a lot less (we work far more often on the base tests than the sample app tests) the weird lines are concentrated in one place, this is inline with the goal of making it simpler for maintainers to work on it I think, what do you think about this?

Reply:
Fair enough